+++
title = "在Ruby中为什么使用`rescue Exception => e`是不好的风格？"
slug = "在Ruby中为什么使用`rescue Exception => e`是不好的风格？"
date = "2023-08-28T06:54:45+08:00"
tags = [ "exception", "ruby",]
thumbnail = "/images/ruby.256x256.png"
featureImage = "/images/ruby.256x256.png"
+++


## 摘要
`rescue Exception => e`会捕获所有异常，这往往不是我们希望的行为。在处理异常时，应该只处理我们知道如何处理的异常。通常情况下，应该使用更具体的异常类进行捕获。如果需要访问异常对象，可以使用`rescue => e`。

## 内容
`Exception`是Ruby异常继承层次结构的根，当使用`rescue Exception`时，实际上是捕获了所有的异常，包括`SyntaxError`、`LoadError`和`Interrupt`等子类。这样的捕获行为可能产生意想不到的结果。

例如，如果捕获了`Interrupt`异常，用户将无法使用<kbd>CTRL</kbd>+<kbd>C</kbd>来中止程序。捕获`SignalException`异常将导致程序无法正确响应信号，除非使用`kill -9`来终止程序。捕获`SyntaxError`意味着`eval`失败时不会触发错误。

以下示例程序展示了捕获`Exception`可能带来的问题：

```
loop do
  begin
    sleep 1
    eval "djsakru3924r9eiuorwju3498 += 5u84fior8u8t4ruyf8ihiure"
  rescue Exception
    puts "I refuse to fail or be stopped!"
  end
end
```

另外值得注意的是，捕获`Exception`并不是默认行为。在没有指定异常类的情况下，`rescue`语句默认会捕获`StandardError`。通常情况下，应该明确指定更具体的异常类，而不是捕获`Exception`。捕获`Exception`会扩大异常范围，可能导致严重的后果，并且使问题排查变得非常困难。

对于需要捕获`StandardError`并且需要访问异常对象的情况，可以使用下面的语法：

```
begin
  # ...
rescue => e
  # ...
end
```

这等效于：

```
begin
  # ...
rescue StandardError => e
  # ...
end
```

为了记录和报告目的而捕获`Exception`是少数几个可以接受的情况之一，在这种情况下，应该立即重新抛出异常，避免潜在的问题。

```
begin
  # ...
rescue Exception => e
  # 记录日志
  raise # 没有足够的救生艇 ;)
end
```

## 总结
捕获`Exception`是一个捕获所有异常的行为，这往往不是我们期望的行为。在处理异常时，应该只处理我们知道如何处理的异常。通常情况下，应该使用更具体的异常类进行捕获。如果需要访问异常对象，可以使用`rescue => e`。使用`rescue Exception`会扩大异常范围，可能导致严重的后果，并使问题排查变得困难。保持代码的可读性和一致性有助于准确定位和解决问题。


